### ISSUE TITLE
Allow setting scale domain for binned field

### ISSUE THREAD

**Issue by mpetris** [vega/vega-lite#2791]
Hi,
I use a bin encoding with 10 steps on a scale range from 0 to 100. Sometimes I have empty bins but I would like to always show the full range [0,100] on the x-axis. The docs say a range specification for the x scale will be ignored. So is there another way?

**Comment by kanitw** [comment]
I guess you mean your scale `domain` from `[0, 100]` (data values ), not range (which represent visual/pixel values).

Currently, please set your `bin`'s `extent` to `[0, 100]`.  

```
encoding: {
   x: {bin: {extent: [0, 100], ...},
  ...
}
```
Let me know if that works for you :)

**Comment by kanitw** [comment]
Hmm actually this turns out to not work like I expected

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"url": "data/cars.json"},
  "mark": "bar",
  "encoding": {
    "x": {
      "bin": {"maxbins": 10, "extent": [0, 30], "nice": false},
      "field": "Acceleration",
      "type": "quantitative"
    },
    "y": {
      "aggregate": "count",
      "type": "quantitative"
    }
  }
}
```

I think we may have to fix by allowing `scale.domain` for binned field.

**Comment by mpetris** [comment]
@kanitw thanks for looking into this so quickly! Alas, it's not working as you said. 
This is stripped down version of my use case: https://gist.github.com/mpetris/b37f196abf84557f31acbc286e38e788

**Comment by kanitw** [comment]
You're welcome.  Btw I forgot to mention that, for now, you may want to edit scale domain in the compiled Vega output, which will allow you to produce the desired output using Vega.

**Comment by mpetris** [comment]
Yes, that workaround did the trick. Thanks!

**Comment by yhoonkim** [comment]
@kanitw , I think it is a vega issue. 
According to `vega.bin` function, bins should be ranged from 0 to 30.
```js
vega.bin({"maxbins":30, "extent":[0,30]}) // It returns {start: 0, stop: 30, step: 5}.
```
But the transformed data looks like this :
![image](https://user-images.githubusercontent.com/6269456/29390789-828f29fc-82a7-11e7-91c1-c95ea8792e87.png)
I think the transformed data should include the below too. 
```json
{"bin_maxbins_10_extent_-5,30_nice_false_Acceleration_start": 0, "bin_maxbins_10_extent_-5,30_nice_false_Acceleration_end": 5, "count_*": 0, "Symbol(vega_id)": 1341}
{"bin_maxbins_10_extent_-5,30_nice_false_Acceleration_start": 25, "bin_maxbins_10_extent_-5,30_nice_false_Acceleration_end": 30, "count_*": 0, "Symbol(vega_id)": 1341}
```
If it return these two records, the current vega-lite spec will work correctly.

**Comment by domoritz** [comment]
@willium Is it a problem that the fields for bins have a `-` in them?

**Comment by willium** [comment]
I will investigate.

**Comment by kanitw** [comment]
@yhoonkim If you are confident that this is a Vega problem, please file a bug. 
Note that if you specify `domain` to `[0, 30]` in the Vega output too, it works fine. 
Thus, I suspect if it is really a Vega bug.  (You can ask Jeff to confirm.) 

If it is not a bug, the important question here we should answer is how to best re-design Vega-Lite to ensure that we show full range:

1) Do we let users set scale `domain` as well?
Or 2) Do we automatically set scale `domain` to the same value as `extent` when `extent` is set?

**Comment by yhoonkim** [comment]
@kanitw 
I like the second way which is easier.
 
If we allow user to set `scale.domain`, it should be ranged inside of `bin.extent`, but it is the same as setting a filter excluding data outside of the domain.

**Comment by kanitw** [comment]
From some more experiment, it's weird when domain is different from extent in Vega.
For example, we get weird empty spaces when  I set extent to `[0, 30]` but domain to `[0, 50]`

![image](https://user-images.githubusercontent.com/111269/29442287-6d71ac6e-8385-11e7-81b2-1ef2672673ce.png)

Thus, I think extent and domain should be the same value.
Therefore, I agree that setting `extent` should automatically set `scale.domain` to the same value and setting `scale.domain` for binned field should remain prohibited.  


An alternative would be remove `extent` and only allow setting `scale.domain`, which will automatically set both to the same value.  But it's kinda weird that one of the bin parameter is outside of the bin object.  (It's also more than just domain as it affects the bin transform that we perform before the scale.) 

Btw, note that the following is not equivalent

> If we allow user to set scale.domain, it should be ranged inside of bin.extent, but it is the same as setting a filter excluding data outside of the domain.

Basically without `clamp`ing the scale, you can get marks that go beyond the plot width.
Like here I set extent to `[0, 30]` and domain to `[0, 20]`

![image](https://user-images.githubusercontent.com/111269/29442461-7816f34e-8386-11e7-95ea-cf84bf4c9542.png)

**Comment by yhoonkim** [comment]
@kanitw Thanks for experiments! 
I'll work on the way of set `scale.domain` to the same value as  `extent`.

**Comment by kanitw** [comment]
@yhoonkim -- I believe this is fixed?

**Comment by mpetris** [comment]
@yhoonkim, yes that is the look I had in mind. Great! Thanks
